### PR TITLE
create user in AD with the same name -4321

### PR DESCRIPTION
**Pull request by valentinabennett**
**Before creating a pull request make sure that:**

- [ x] commit messages are meaningful and follow good commit message guidelines
- [ x] README and other documentation has been updated / added (if needed)
- [x ] tests have been updated / new tests has been added (if needed)

### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/VIH-4321

### Change description ###
Create user with the same first name and last name in AD by adding suffice

**Does this PR introduce a breaking change?** (check one with "x")

```
[ X] Yes
[ ] No
```

### CODE REVIEW

File: UserApi/UserApi/Services/UserAccountService.cs [hunk 1]
```diff
@@ -279,6 +280,55 @@ public async Task<List<Group>> GetGroupsForUser(string userId)
...
+        /// <summary>
+        /// Determine the next available username for a participant based on username format [firstname].[lastname]
+        /// </summary>
+        /// <param name="firstName"></param>
+        /// <param name="lastName"></param>
+        /// <returns>next available user principal name</returns>
+        public virtual string CheckForNextAvailableUsername(string firstName, string lastName)
+        {
+            var baseUsername = $"{firstName}.{lastName}".ToLower();
+            var userFilter = $@"startswith(userPrincipalName,'{baseUsername}')";
```
Comment (ddikman) [inline UserApi/UserApi/Services/UserAccountService.cs]: Do you need both `$` and `@`?

File: UserApi/UserApi/Services/UserAccountService.cs [hunk 2]
```diff
@@ -279,6 +280,55 @@ public async Task<List<Group>> GetGroupsForUser(string userId)
...
+            var users = QueryUsers(userFilter).ToList();
+            var domain = "@hearings.reform.hmcts.net";
+            if (!users.Any())
+            {
+                var userPrincipalName = $"{baseUsername}{domain}";
```
Comment (ddikman) [inline UserApi/UserApi/Services/UserAccountService.cs]: you can do single line return here `return $"{baseUsername}{domain}";`

File: UserApi/UserApi/Services/UserAccountService.cs [hunk 3]
```diff
@@ -279,6 +280,55 @@ public async Task<List<Group>> GetGroupsForUser(string userId)
...
+        /// Query Graph for users with the same first and last names.
+        /// </summary>
+        /// <param name="filter">The filter</param>
+        /// <returns>List of users</returns>
+        public IList<User> QueryUsers(string filter)
+        {
+            var accessToken = _tokenProvider.GetClientAccessToken(_azureAdConfiguration.ClientId,
+                 _azureAdConfiguration.ClientSecret, "https://graph.windows.net/");
+            var queryUrl = $"https://graph.windows.net/{_azureAdConfiguration.TenantId}/users?$filter={filter}&api-version=1.6";
+            using (var client = new HttpClient())
```
Comment (ddikman) [inline UserApi/UserApi/Services/UserAccountService.cs]: I think it's time we refactor and break the use of http client out to a helper here.

File: UserApi/UserApi/Services/UserAccountService.cs (hunk 1 above)
Comment (valentinabennett) [reply UserApi/UserApi/Services/UserAccountService.cs]: Removed from several places, as we need only $

File: UserApi/UserApi/Services/UserAccountService.cs (hunk 2 above)
Comment (valentinabennett) [reply UserApi/UserApi/Services/UserAccountService.cs]: Done

File: UserApi/UserApi/Helper/HttpClientHelper.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
+﻿using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Threading.Tasks;
+
+namespace UserApi.Helper
+{
+    public interface IHttpClientHelper
+    {
+        HttpResponseMessage CreateHttpClientGet(string accessToken, string accessUri);
```
Comment (ddikman) [inline UserApi/UserApi/Helper/HttpClientHelper.cs]: Just calling it `Get` would be enough no?

File: UserApi/UserApi/Helper/HttpClientHelper.cs [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+        Task<HttpResponseMessage> CreateHttpClientGetAsync(string accessToken, string accessUri);
+
+        Task<HttpResponseMessage> CreateHttpClientPatchOrPostAsync(string accessToken, StringContent stringContent, string accessUri, HttpMethod httpMethod);
+    }
+
+    public class HttpClientHelper : IHttpClientHelper
+    {
+        public HttpResponseMessage CreateHttpClientGet(string accessToken, string accessUri)
```
Comment (ddikman) [inline UserApi/UserApi/Helper/HttpClientHelper.cs]: Would it be possible to return a task here instead so we don't block the thread on this?

**Review by ddikman — APPROVED** [review]

**Pull request merged**